ntoskrnl/io/iomgr/deviface.c

```diff
@@ -412,6 +466,137 @@ IopOpenInterfaceKey(IN CONST GUID *InterfaceClassGuid,
     return Status;
 }
 
+/*++
+ * @name IoGetDeviceInterfaceAlias
+ * @implemented
+ *
+ * Returns the alias device interface of the specified device interface
+ * instance, if the alias exists.
+ * Documented in WDK.
+ *
+ * @param SymbolicLinkName
+ *        Pointer to a string which identifies the device interface instance
+ *
+ * @param AliasInterfaceClassGuid
+ *        See WDK
+ *
+ * @param AliasSymbolicLinkName
+ *        See WDK
+ *
+ * @return Three different NTSTATUS values in case of errors, and STATUS_SUCCESS
+ *         otherwise (see WDK for details)
+ *
+ * @remarks Must be called at IRQL = PASSIVE_LEVEL in the context of a system thread
+ *
+ *--*/
+NTSTATUS
+NTAPI
+IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+                          IN CONST GUID *AliasInterfaceClassGuid,
+                          OUT PUNICODE_STRING AliasSymbolicLinkName)
+{
+    UNICODE_STRING AliasSymbolicLink = {0, 0, NULL};
+    UNICODE_STRING DosDevicesPrefix = RTL_CONSTANT_STRING(L"\\??\\");
+    UNICODE_STRING GuidString, AliasGuidString;
+    UNICODE_STRING DeviceString, ReferenceString;
+    HANDLE InterfaceKey = NULL;
+    NTSTATUS Status;
+
+    DeviceString.Buffer = ReferenceString.Buffer = GuidString.Buffer = AliasGuidString.Buffer = NULL;
+
+    DPRINT1("IoGetDeviceInterfaceAlias() : symbolic link '%wZ'\n", SymbolicLinkName);
+
+    /* Sanity check */
+    if (SymbolicLinkName == NULL || AliasInterfaceClassGuid == NULL)
+    {
+        DPRINT1("IoGetDeviceInterfaceAlias() : invalid symbolic link or alias class GUID\n");
+        return STATUS_INVALID_HANDLE;
```

Comment:
Are you sure about this value?

Reply:
:thinking: I doubt: in my opinion `STATUS_INVALID_PARAMETER` is more suitable in this case, but MSDN says that if any of this parameters is invalid, then the function should necessarily return `STATUS_INVALID_HANDLE`. So I've just follow the official documentation.

Reply:
It returns `STATUS_INVALID_HANDLE` because it tries to open their related registry keys for verifying they are describing an actual thing.

Reply:
OK, so as I understand, here I should return `STATUS_INVALID_PARAMETER`. Then `STATUS_INVALID_HANDLE` is used when failed to open an appropriate alias subkey.

Reply:
Just write a test that will tell you what actually happens.

Reply:
I already wrote the test for that in my #3595 PR. Just checked: on XP, when the subkey does not contain the needed values, `IoGetDeviceInterfaceAlias` returns `STATUS_OBJECT_NAME_NOT_FOUND`:

![kmtest_result](https://user-images.githubusercontent.com/26385117/123708154-9a824780-d873-11eb-9fd5-2da1d937d507.png)
Same as described in MSDN btw.
But I didn't add (so didn't check) invalid parameters case in my test yet.

Reply:
> IoGetDeviceInterfaceAlias returns STATUS_OBJECT_NAME_NOT_FOUND

Ah, so it's not even `STATUS_INVALID_HANDLE`

Reply:
OK, just added an invalid parameter testcases and tested with my updated kmtest. :slightly_smiling_face: 
At least when I'm passing an invalid (`NULL`) symbolic link, it returns `STATUS_INVALID_PARAMETER`, not `STATUS_INVALID_HANDLE`. With invalid GUID the same value should be returned I guess, because MSDN says that it returns identical value if one of both parameters is invalid. So I should return the same in my implementation.
![invalid_parameter](https://user-images.githubusercontent.com/26385117/149626350-c0f6c633-90c2-430e-9be3-55764074d214.png)
But I still don't understand why MSDN lies about the return value itself. Maybe it has been changed if the future Windows versions, but why then there is no any note about that then?... :thinking: Although these are related return codes, but nevertheless...

Reply:
Fixed return value. :ok_hand: